Get the part builder up and running

alexbegt opened the pull request:
What will we do without the part builder?

alexbegt committed:
Get the part builder up and running!

KnightMiner committed:
Implement pattern icons in part builder GUI

TODO: would like to make it dynamic if possible, but cannot find a way yet

KnightMiner committed:
Remove unneeded deferred wrapper

It did nothing more than wrap the output, so might as well use the original

alexbegt committed:
Add support for showing the material stats in the part builder

alexbegt committed:
Fix the side info panel's scroll bar not functioning.

alexbegt committed:
Fix the Info Panel claiming the mouse click even if the user clicks nowhere near the scroll bar.

alexbegt committed:
Make getLocalizedDesc return ITextComponent instead of string

alexbegt committed:
Fix the CustomFontRenderer including the custom font color's in it's character width calculation and fix the spacing between characters.

alexbegt committed:
Add getting the name of the current tool's harvest level; TODO: Move calculation out of 'static' and into a new home. (Was in Tconstruct.preInit in 1.12)

alexbegt committed:
Add the wide guard, shovel head, and sword blade tool parts for future use and switch to using ToolPart instead of MaterialItem for parts

alexbegt committed:
Fix the part chest not scrolling the side bars

alexbegt committed:
Fix the config option being wrong for tool parts

alexbegt committed:
Fix shovel head placement

alexbegt committed:
Add the large plate, hammer head and tough tool rod

alexbegt committed:
Fix all the errors I created that stops the part builder branch from working

alexbegt committed:
Fix the missing colors on the new tool parts

alexbegt committed:
Move some textures around for the parts I've added, which ironically includes the tools

alexbegt committed:
Fix all the missing names and tool tips

alexbegt committed:
Remove the unneeded tool station loot table

alexbegt committed:
Fix the location of the shovel head in the part builder

KnightMiner commented on src/main/java/slimeknights/tconstruct/TConstruct.java:
```diff
@@ -118,6 +118,7 @@ public TConstruct() {
 
   @SubscribeEvent
   public static void commonSetup(final FMLCommonSetupEvent event) {
+    DistExecutor.runWhenOn(Dist.CLIENT, () -> TinkerClient::onCommonSetup);
```
Don't do this here, make it an event handler in TinkerClient. I want to keep `DistExecutor` calls to a minimum.

KnightMiner commented on src/main/java/slimeknights/tconstruct/common/config/Config.java:
```diff
@@ -39,7 +39,9 @@
-    public final ForgeConfigSpec.BooleanValue allowStencilsToBeReused;
+    public final ForgeConfigSpec.BooleanValue listAllToolMaterials;
+
+    public final ForgeConfigSpec.BooleanValue listAllPartMaterials;
```
I feel like these should be client config options. On the server, always use the full list.

KnightMiner commented on src/main/java/slimeknights/tconstruct/common/data/TConstructRecipeProvider.java:
```diff
@@ -692,6 +700,145 @@ private void addSlimeRecipes(Consumer<IFinishedRecipe> consumer) {
+      .addCriterion("has_item", this.hasItem(TinkerTables.pattern.get()))
+      .build(consumer, location("parts/small_binding"));
+
+    PartRecipeBuilder.partRecipe(new ItemStack(TinkerToolParts.wideGuard.get()))
```
Might be worth a local helper for these, as the pattern is always the register name, and the criteria is always `this.hasItem(pattern)`. Something like `addPartRecipe(IItemProvider part, int cost)`

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/client/Icons.java:
```diff
@@ -6,23 +6,23 @@
-  ElementScreen ICON_Button = new ElementScreen(180, 216, 18, 18);
-  ElementScreen ICON_ButtonHover = new ElementScreen(180 + 18 * 2, 216, 18, 18);
-  ElementScreen ICON_ButtonPressed = new ElementScreen(180 - 18 * 2, 216, 18, 18);
+  ElementScreen ANVIL_ICON = new ElementScreen(18 * 3, 0, 18, 18, 256, 256);
```
Suffixing everything with `_ICON` is redundant, the class is named `Icons`

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/materials/stats/BaseMaterialStats.java:
```diff
@@ -37,15 +23,15 @@ public static String formatNumber(String loc, String color, int number) {
 
   public static String formatNumber(String loc, String color, float number) {
```
Should return `ITextComponent`, though we can put that off to a later commit.

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/materials/stats/BaseMaterialStats.java:
```diff
@@ -15,20 +14,7 @@
 
   @Override
   public String getLocalizedName() {
```
Should return `ITextComponent`

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/materials/stats/BaseMaterialStats.java:
```diff
@@ -15,20 +14,7 @@
-  public List<String> getLocalizedDesc() {
-    // todo
-    return null;
+    return String.format("stat.%s.name", this.getIdentifier().getPath());
```
Isn't this supposed to be localized? If going unlocalized, call it `getTranslationKey`

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/tinkering/ToolPart.java:
```diff
@@ -0,0 +1,168 @@
+  @Override
+  public boolean hasUseForStat(MaterialStatsId stat) {
+    //TODO check if part is correct for any tool's required components
+    return true;
```
Isn't this just an equals check with the part's stat ID?

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/tinkering/ToolPart.java:
```diff
@@ -0,0 +1,168 @@
+    return builder.build();
+  }
+
+  public List<ITextComponent> getAddedByInfo(IMaterial material) {
```
Materials are now namespaced, this should be unneeded.

alexbegt replied:
I left this for now, can be removed later

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/tools/helper/ToolAttackUtil.java:
```diff
@@ -0,0 +1,54 @@
+import slimeknights.tconstruct.library.tools.ToolCore;
+import slimeknights.tconstruct.library.tools.nbt.ToolData;
+
+public class ToolAttackUtil {
```
`ToolAttackUtil` seems very narrow in focus, didn't we have a general tool helper util in 1.12?

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/tools/helper/ToolAttackUtil.java:
```diff
@@ -0,0 +1,54 @@
+    return damage;
+  }
+
+  public static float calcCutoffDamage(float damage, float cutoff) {
```
Javadocs for new methods, especially helpers

KnightMiner commented on src/main/java/slimeknights/tconstruct/shared/TinkerClient.java:
```diff
@@ -39,4 +42,17 @@ private static void addResourceListeners(IReloadableResourceManager manager) {
+    if (minecraft != null) {
+      IResourceManager manager = Minecraft.getInstance().getResourceManager();
+      if (manager instanceof IReloadableResourceManager) {
+        ((IReloadableResourceManager) manager).addReloadListener(fontRenderer);
```
Use existing method for this, that is the whole reason I added `addResourceListeners`

alexbegt replied:
I am unable to use `addResourceListeners` due to it being called before the font renderer has been constructed, which leads to a NPE crash

KnightMiner commented on src/main/java/slimeknights/tconstruct/shared/TinkerClient.java:
```diff
@@ -39,4 +42,17 @@ private static void addResourceListeners(IReloadableResourceManager manager) {
     manager.addReloadListener(textureValidator);
   }
+
+  public static void onCommonSetup() {
```
Subscribe to the event for this. It possibly should be moved to shared client events, maybe I should just merge the two

KnightMiner commented on src/main/java/slimeknights/tconstruct/tables/client/inventory/module/InfoPanelScreen.java:
```diff
@@ -198,17 +193,17 @@ protected void updateSliderParameters() {
+      this.tooltipLines.add(lines.size());
       // empty line
-      if (line == null || line.isEmpty()) {
+      if (textComponent == null || textComponent.getFormattedText().isEmpty()) {
```
Store formatted text in a variable and reuse below, in case its inefficient.

KnightMiner commented on src/main/java/slimeknights/tconstruct/tables/inventory/table/PartBuilderContainer.java:
```diff
@@ -148,4 +155,71 @@ public void onCrafting(PlayerEntity playerEntity, ItemStack output, IInventory c
+  }
+
+  public PartRecipe getPartRecipe() {
+    return this.partRecipes.get(this.selectedRecipe.get());
```
I'd prefer to fetch the part recipe with a matches clause. The interface it gets should contain the selected pattern and the selected material. Lets us have multiple recipes for one pattern.

KnightMiner replied:
Going to leave this one for now. I am considering a few different options for a way to do custom pattern recipes, but its not a priority

KnightMiner commented on src/main/java/slimeknights/tconstruct/tables/inventory/table/PartBuilderContainer.java:
```diff
@@ -148,4 +155,71 @@ public void onCrafting(PlayerEntity playerEntity, ItemStack output, IInventory c
+    this.craftResult.setInventorySlotContents(0, ItemStack.EMPTY);
+
+    if (!stack.isEmpty()) {
+      this.partRecipes = this.world.getRecipeManager().getRecipes(TinkerTables.partRecipeType, inventoryIn, this.world);
```
Instead of each part recipe being a button, I would rather patterns be separate from part recipes. So basically, iterate all recipes to get a list of unique patterns. This will allow us to add non-tool part recipes later, such as an ingot pattern (mostly for modpacks).

We could work on this in a later commit instead of here.

KnightMiner commented on src/main/java/slimeknights/tconstruct/tables/inventory/table/PartBuilderContainer.java:
```diff
@@ -65,80 +85,67 @@ public PartBuilderContainer(int id, PlayerInventory inv, PacketBuffer buf) {
 
   @Override
   public void onCraftMatrixChanged(IInventory inventoryIn) {
+    ItemStack itemstack = this.patternSlot.getStack();
```
Store material recipe into a variable, then let the interface grab it in the recipe.

KnightMiner commented on src/main/java/slimeknights/tconstruct/tables/recipe/material/MaterialRecipe.java:
```diff
@@ -0,0 +1,122 @@
+   * Used to check if a recipe matches current crafting inventory
+   */
+  @Override
+  public boolean matches(IInventory inv, World worldIn) {
```
May need to check multiple slots when used in the tool station, though we can rethink that later when needed.

KnightMiner commented on src/main/java/slimeknights/tconstruct/tables/recipe/material/MaterialRecipe.java:
```diff
@@ -0,0 +1,122 @@
+    return true;
+  }
+
+  /**
+   * Get the result of this recipe, usually for display purposes (e.g. recipe book). If your recipe has more than one
+   * possible result (e.g. it's dynamic and depends on its inputs), then return an empty stack.
+   */
```
Lets skip copying vanilla javadocs

KnightMiner commented on src/main/java/slimeknights/tconstruct/tables/recipe/material/MaterialRecipe.java:
```diff
@@ -0,0 +1,122 @@
+  protected final Ingredient ingredient;
+  protected final MaterialId materialId;
+  protected final int value;
+  protected final int needed;
```
Lombok will make you getters if you want

KnightMiner commented on src/main/java/slimeknights/tconstruct/tables/recipe/part/PartRecipe.java:
```diff
@@ -0,0 +1,99 @@
+   */
+  @Override
+  public boolean matches(IInventory inv, World worldIn) {
+    return inv.getStackInSlot(1).getItem() == TinkerTables.pattern.get();
```
I don't like the hardcoded slot here, should have the interface grant only the pattern slot. Plus the matches should really check the selected pattern button as another interface return, along with check if the passed material is valid for this part.

KnightMiner commented on src/main/java/slimeknights/tconstruct/tables/recipe/part/PartRecipe.java:
```diff
@@ -0,0 +1,99 @@
+   */
+  @Override
+  public ItemStack getCraftingResult(IInventory inv) {
+    return this.output.copy();
```
Set material for the output here, based on the material returned from the interface

KnightMiner commented on src/main/java/slimeknights/tconstruct/tables/recipe/part/PartRecipe.java:
```diff
@@ -0,0 +1,99 @@
+import net.minecraft.world.World;
+import slimeknights.tconstruct.tables.TinkerTables;
+
+public class PartRecipe implements IRecipe<IInventory> {
```
Extend `IInventory` into an interface that better suits the part builder. Should contain methods such as `getInput` (item in material slot), `getMaterial` (computed from the input), and `getPattern` (selected pattern button resource location). Maybe also have a `getPatternItem` for that slot instead of the hardcoded index. Have the interface default anything we don't care about, maybe pulling those defaults to a base `IInventoryHasToManyMethods`

KnightMiner commented on src/main/java/slimeknights/tconstruct/tables/tileentity/table/PartBuilderTileEntity.java:
```diff
@@ -14,7 +14,7 @@
 
   public PartBuilderTileEntity() {
-    super(TinkerTables.partBuilderTile.get(), "gui.tconstruct.part_builder", 4);
+    super(TinkerTables.partBuilderTile.get(), "gui.tconstruct.part_builder", 3);
```
What is that magic number at the end? Should likely move to a constant somewhere

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/tinkering/ToolPart.java:
```diff
@@ -0,0 +1,168 @@
+import java.util.List;
+import java.util.Optional;
+
+public class ToolPart extends MaterialItem implements IToolPart {
```
`ToolPartItem`

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/tinkering/ToolPart.java:
```diff
@@ -0,0 +1,168 @@
+  }
+
+  @Override
+  public boolean canUseMaterial(IMaterial material) {
```
I would also put an overridable `canUseMaterial` over in `MaterialItem`. Might let us move the sub items code there, but mostly because that concept is relevant to all material items.

KnightMiner reviewed (commented):
Added some general comments. A few we can deal with later. One thing it would not let me comment on was I see you have two copies of the GUI texture for large plates, you can delete the one from the parts folder now.

alexbegt committed:
Move the CustomFontRenderer to the CommonsClientEvents instead of using a runWhenOn

alexbegt committed:
Rename ToolPart -> ToolPartItem

alexbegt committed:
Add helper method for registering part recipes

alexbegt committed:
Switch to using ITextComponets instead of strings

alexbegt committed:
Fix some missing language issues

alexbegt committed:
Fix NPE issues with getting the Text Component for the diamond level pickaxes.

alexbegt committed:
Fix there being tooltips that render as blank strings.

alexbegt committed:
Add some javadoc to ToolAttackUtil

alexbegt committed:
Merge branch '1.15' into partbuilder

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/materials/stats/BaseMaterialStats.java:
```diff
@@ -14,40 +15,25 @@
-      TextFormatting.RESET);
+  public static ITextComponent formatNumber(String loc, String color, float number) {
+    return new TranslationTextComponent(loc)
+      .appendSibling(new StringTextComponent(": "))
```
Might as well use append string here. My comment on Discord was if we do it using appendSibling, we can then format the text component we pass into appendSibling:

```java
return new TranslationTextComponent(loc)
    .appendSibling(new StringTextComponent("whatever").applyTextStyle(TextFormatting.WHITE));
```

Since we could not get color working in 1.15, might as well stick to `appendString` for now.

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/tinkering/ToolPartItem.java:
```diff
@@ -0,0 +1,170 @@
+
+  @Override
+  public boolean hasUseForStat(MaterialStatsId stat) {
+    //TODO check if part is correct for any tool's required components
```
Probably should be renamed (`isStatType`?), but the code is literally `return this.materialStatId.equals(stat)`

KnightMiner commented on src/main/java/slimeknights/tconstruct/library/tinkering/ToolPartItem.java:
```diff
@@ -0,0 +1,170 @@
+      tooltip.add(error.getFormattedText());
+      return true;
+    }
+    /*else if(statIdentifier != null && material.getStats(statIdentifier) == null) {
```
Why is this commented out? It looks like it would be useful

alexbegt replied:
`getStats(statIdentifier)` is not a method on IMaterial, so I commented it out

KnightMiner commented on src/main/java/slimeknights/tconstruct/shared/TinkerClient.java:
```diff
@@ -3,10 +3,16 @@
 import net.minecraft.client.Minecraft;
 import net.minecraft.resources.IReloadableResourceManager;
 import net.minecraft.resources.IResourceManager;
+import net.minecraftforge.api.distmarker.Dist;
+import net.minecraftforge.api.distmarker.OnlyIn;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.event.lifecycle.FMLCommonSetupEvent;
 import net.minecraftforge.resource.VanillaResourceType;
 import slimeknights.tconstruct.library.book.TinkerBook;
 import slimeknights.tconstruct.library.client.materials.MaterialRenderInfoLoader;
+import slimeknights.tconstruct.library.client.renderer.font.CustomFontRenderer;
 import slimeknights.tconstruct.library.client.util.ResourceValidator;
+import slimeknights.tconstruct.tools.ToolClientEvents;
```
These imports are probably no longer needed

KnightMiner reviewed (approved):
Left a few comments, minor style stuff and unused code, but one function that should be implemented (its trivial). Should probably be good to merge and we can do some cleanups later if needed.

alexbegt committed:
Clean up unneeded imports

alexbegt committed:
Enable the trait tooltip (showes please implement methodname), implement hasUseForStat and fix up checkMissingMaterialTooltip.

alexbegt committed:
Clean up the ITextComponents used for stats and switch to appendText for now.

KnightMiner committed:
Move canUseMaterial to IMaterialItem

Allows the part builder to be more flexible as you do not have to craft tool parts

KnightMiner committed:
Repackage material and part recipes into library

Also move the recipe types

KnightMiner committed:
Clean up tool part items

Only a single stat type per item now

KnightMiner committed:
Make part builder generic to interfaces instead of our specific items

KnightMiner committed:
Move most part builder recipe logic into the part recipe

Lets us remove the tool builder helper and make more powerful recipes

KnightMiner committed:
Clean up part builder UI rendering code

KnightMiner committed:
Use textures instead of models for the part textures

Lets me automatically load them in, plus models were overkill.
From this point on, any texture in the folder gui/tinker_pattern is automatically stitched

KnightMiner committed:
Force part recipes to output a material item

If we want one that outputs a regular item, can make a separate recipe later. Current UI structure prevents that
